### ISSUE TITLE
Feature request: add FlareSolverr solver to bypass CloudFlare protection

### ISSUE THREAD

**Issue by soredake** [lycheeverse/lychee#1439]
Ther are a lot of sites that use cloudflare as protection, checking sites that use cloudflare can result in 403, would be nice to have support for [FlareSolverr ](https://github.com/FlareSolverr/FlareSolverr) to fix this situation.

Related:
https://github.com/lycheeverse/lychee/issues/1157
https://github.com/jobobby04/TachiyomiSY/pull/1124
https://github.com/Suwayomi/Suwayomi-VaadinUI/commit/712ca2a86b9d853a7e0dfcc067b72c00cb399844

**Comment by mre** [comment]
First time I hear of FlareSolverr, but it looks nice. We wanted to add proxy support anyway at some point. There's a proxy config file that browsers like Firefox use. Not sure if it could also be used to call FlareSolverr.

**Comment by mre** [comment]
The proxy support request I was referring to: https://github.com/lycheeverse/lychee/issues/869
Would it "just work" if we added support for that?

**Comment by mre** [comment]
Looks like `FlareSolverr` requires data to be sent as a JSON `POST` request? So the proxy approach wouldn't work.
There's a prototype for WebAssembly support, which could be used to modify the request. @thomas-zahner, this should work, right?

**Comment by thomas-zahner** [comment]
Yes this seems like a perfect use case of the recently implemented [`RequestChain`](https://docs.rs/lychee-lib/latest/lychee_lib/chain/index.html). The WebAssembly part is indeed only a prototype and not merged yet. However, the request chain itself is already fully functional and available in `lychee_lib` through the `ClientBuilder`.

An example on how to make use of the `RequestChain` can be found here: https://github.com/lycheeverse/lychee/blob/master/examples/chain/chain.rs

**Comment by soredake** [comment]
Another example https://github.com/Suwayomi/Suwayomi-VaadinUI/commit/712ca2a86b9d853a7e0dfcc067b72c00cb399844

**Comment by thomas-zahner** [comment]
So I've created an example on how to use the `RequestChain` to call the FlareSolverr proxy instead of the actual website and then use the status code returned inside FlareSolverr's response body. Here you can find the branch: https://github.com/thomas-zahner/lychee/commit/297c2e93dfab3e46985bc34146ca03c67d8513df

Note that I didn't test it on a protected website. But as long as the status code returned by FlareSolverr is accurate it should work as expected. I felt that FlareSolverr's documentation is not very good as I didn't see the "official" JSON response schema, so I've used their provided "example response" which hopefully should represent the consistent JSON response schema.

**Comment by thomas-zahner** [comment]
Noticed a bug in my implementation, currently `Status:Ok` is always returned. Fix would be a little hacky (copy paste) so I've [opened up an issue](https://github.com/lycheeverse/lychee/issues/1472)

**Comment by MHammett** [comment]
A data point in favour of a lighter option than a FlareSolverr sidecar, if it's useful.

I measured six real Cloudflare-blocked URLs from a citation-checking workload on 2026-08-12:

- **Full browser header set** (UA, Accept, Accept-Language, Sec-Fetch-*): **0/6**. The hard blocks fingerprint the TLS handshake, so no header string moves them — they 403 a plain request to the domain root just the same.
- **Browser TLS fingerprint** (`curl_cffi`, `impersonate="chrome"`): **2/5** recovered with real content — congress.gov (488 KB) and a CDC PDF (661 KB).

The remaining three (ASME, Wiley/AGU, Royal Society) returned a ~6 KB challenge page to everything I tried. I'm fairly confident those are subscription gates rather than bot gates, and no client-side fix reaches them.

The takeaway: matching the TLS fingerprint recovers a meaningful slice of what a headless-browser solver would, without standing up a sidecar service. The equivalent knob on the Rust side is a TLS backend that can present a browser ClientHello rather than reqwest's default — which is a much smaller dependency question than FlareSolverr, though I'll leave the specifics to people who know lychee's internals.

Worth gating behind an explicit opt-in either way. Impersonating a browser should be a deliberate choice a user makes, not something a link checker does by default.
